File: billing-v3/src/main/java/io/appmetrica/analytics/billingv3/impl/update/UpdatePolicyImpl.java
package io.appmetrica.analytics.billingv3.impl.update;

import androidx.annotation.NonNull;
import io.appmetrica.analytics.billinginterface.internal.BillingInfo;
import io.appmetrica.analytics.billinginterface.internal.ProductType;
import io.appmetrica.analytics.billinginterface.internal.config.BillingConfig;
import io.appmetrica.analytics.billinginterface.internal.storage.BillingInfoManager;
import io.appmetrica.analytics.billinginterface.internal.update.UpdatePolicy;
import io.appmetrica.analytics.coreutils.internal.logger.YLogger;
import io.appmetrica.analytics.coreutils.internal.time.SystemTimeProvider;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class UpdatePolicyImpl implements UpdatePolicy {

    private static final String TAG = "[UpdatePolicyImpl]";

    @NonNull
    private final SystemTimeProvider systemTimeProvider;

    public UpdatePolicyImpl() {
        this(new SystemTimeProvider());
    }

    UpdatePolicyImpl(@NonNull final SystemTimeProvider systemTimeProvider) {
        this.systemTimeProvider = systemTimeProvider;
    }

    @NonNull
    @Override
    public Map<String, BillingInfo> getBillingInfoToUpdate(@NonNull final BillingConfig config,
                                                           @NonNull final Map<String, BillingInfo> history,
                                                           @NonNull final BillingInfoManager storage) {
        YLogger.info(TAG, "getNewBillingInfo");
        final Map<String, BillingInfo> newBillingInfo = new HashMap<>();
        for (final String sku: history.keySet()) {
            final BillingInfo historyEntry = history.get(sku);
            if (shouldUpdateBillingInfo(config, historyEntry, storage)) {
                YLogger.info(TAG, "Product %s should be updated", historyEntry.sku);
                newBillingInfo.put(sku, historyEntry);
            } else {
                YLogger.info(TAG, "Product %s should be ignored", historyEntry.sku);
            }
        }
        return newBillingInfo;
    }

    private boolean shouldUpdateBillingInfo(@NonNull final BillingConfig config,
                                            @NonNull final BillingInfo historyEntry,
                                            @NonNull final BillingInfoManager storage) {
        final long now = systemTimeProvider.currentTimeMillis();
        YLogger.info(TAG, "isFirstInappCheckOccurred " + storage.isFirstInappCheckOccurred());
        if (historyEntry.type == ProductType.INAPP && !storage.isFirstInappCheckOccurred()) {
            return now - historyEntry.purchaseTime <=
                    TimeUnit.SECONDS.toMillis(config.firstCollectingInappMaxAgeSeconds);
        }
        final BillingInfo storageEntry = storage.get(historyEntry.sku);
        if (storageEntry == null) {
            return true;
        }
        if (!storageEntry.purchaseToken.equals(historyEntry.purchaseToken)) {
            return true;
        }
        if (historyEntry.type == ProductType.SUBS) {
            return now - storageEntry.sendTime >= TimeUnit.SECONDS.toMillis(config.sendFrequencySeconds);
        }
        return false;
    }
}

File: billing-v4/src/main/java/io/appmetrica/analytics/billingv4/impl/update/UpdatePolicyImpl.kt
package io.appmetrica.analytics.billingv4.impl.update

import io.appmetrica.analytics.billinginterface.internal.BillingInfo
import io.appmetrica.analytics.billinginterface.internal.ProductType
import io.appmetrica.analytics.billinginterface.internal.config.BillingConfig
import io.appmetrica.analytics.billinginterface.internal.storage.BillingInfoManager
import io.appmetrica.analytics.billinginterface.internal.update.UpdatePolicy
import io.appmetrica.analytics.coreutils.internal.logger.YLogger
import io.appmetrica.analytics.coreutils.internal.time.SystemTimeProvider
import java.util.concurrent.TimeUnit

private const val TAG = "[UpdatePolicyImpl]"

class UpdatePolicyImpl(
    private val systemTimeProvider: SystemTimeProvider = SystemTimeProvider()
) : UpdatePolicy {

    override fun getBillingInfoToUpdate(
        config: BillingConfig,
        history: Map<String, BillingInfo>,
        storage: BillingInfoManager
    ): Map<String, BillingInfo> {
        val productsToUpdate = history.filterValues {
            shouldUpdateBillingInfo(config, it, storage)
        }
        YLogger.info(TAG, "Products ${productsToUpdate.values.map { it.sku }} should be updated")
        return productsToUpdate
    }

    private fun shouldUpdateBillingInfo(
        config: BillingConfig,
        historyEntry: BillingInfo,
        storage: BillingInfoManager
    ): Boolean {
        val now = systemTimeProvider.currentTimeMillis()
        if (historyEntry.type == ProductType.INAPP && !storage.isFirstInappCheckOccurred) {
            return now - historyEntry.purchaseTime <=
                TimeUnit.SECONDS.toMillis(config.firstCollectingInappMaxAgeSeconds.toLong())
        }
        val storageEntry = storage[historyEntry.sku] ?: return true
        if (storageEntry.purchaseToken != historyEntry.purchaseToken) {
            return true
        }
        return if (historyEntry.type == ProductType.SUBS) {
            now - storageEntry.sendTime >= TimeUnit.SECONDS.toMillis(config.sendFrequencySeconds.toLong())
        } else false
    }
}
